runtime/mux_test.go

```diff
@@ -466,3 +475,126 @@ func TestServeMux_HandlePath(t *testing.T) {
 		})
 	}
 }
+
+var healthCheckTests = []struct {
+	name           string
+	code           codes.Code
+	status         grpc_health_v1.HealthCheckResponse_ServingStatus
+	httpStatusCode int
+}{
+	{
+		"Test grpc error code",
+		codes.NotFound,
+		grpc_health_v1.HealthCheckResponse_UNKNOWN,
+		http.StatusNotFound,
+	},
+	{
+		"Test HealthCheckResponse_SERVING",
+		codes.OK,
+		grpc_health_v1.HealthCheckResponse_SERVING,
+		http.StatusOK,
+	},
+	{
+		"Test HealthCheckResponse_NOT_SERVING",
+		codes.OK,
+		grpc_health_v1.HealthCheckResponse_NOT_SERVING,
+		http.StatusServiceUnavailable,
+	},
+	{
+		"Test HealthCheckResponse_UNKNOWN",
+		codes.OK,
+		grpc_health_v1.HealthCheckResponse_UNKNOWN,
+		http.StatusServiceUnavailable,
+	},
+	{
+		"Test HealthCheckResponse_SERVICE_UNKNOWN",
+		codes.OK,
+		grpc_health_v1.HealthCheckResponse_SERVICE_UNKNOWN,
+		http.StatusNotFound,
+	},
+}
+
+func TestWithHealthzEndpoint_codes(t *testing.T) {
+	for _, tt := range healthCheckTests {
+		t.Run(tt.name, func(t *testing.T) {
+			conn, err := dummyGrpcServer(tt.status, tt.code)
+			if err != nil {
+				t.Errorf("connection to dummy grpc server failed: %v", err)
+			}
+
+			mux := runtime.NewServeMux(runtime.WithHealthzEndpoint(conn))
+
+			r := httptest.NewRequest(http.MethodGet, "/healthz", nil)
+			rr := httptest.NewRecorder()
+
+			mux.ServeHTTP(rr, r)
+
+			if rr.Code != tt.httpStatusCode {
+				t.Errorf(
+					"result http status code for grpc code %q and status %q should be %d, got %d",
+					tt.code, tt.status, tt.httpStatusCode, rr.Code,
+				)
+			}
+		})
+	}
+}
+
+func TestWithHealthzEndpoint_serviceParam(t *testing.T) {
+	service := "test"
+
+	// set codes.UNKNOWN to trigger error response with service in msg
+	conn, err := dummyGrpcServer(grpc_health_v1.HealthCheckResponse_UNKNOWN, codes.Unknown)
+	if err != nil {
+		t.Errorf("connection to dummy grpc server failed: %v", err)
+	}
+
+	mux := runtime.NewServeMux(runtime.WithHealthzEndpoint(conn))
+
+	r := httptest.NewRequest(http.MethodGet, "/healthz?service="+service, nil)
+	rr := httptest.NewRecorder()
+
+	mux.ServeHTTP(rr, r)
+
+	if !strings.Contains(rr.Body.String(), service) {
+		t.Errorf(
+			"service query parameter should be translated to HealthCheckRequest: expected %s to contain %s",
+			rr.Body.String(), service,
+		)
+	}
+}
+
+type grpcServer struct {
+	status grpc_health_v1.HealthCheckResponse_ServingStatus
+	code   codes.Code
+}
+
+func (g *grpcServer) Check(ctx context.Context, r *grpc_health_v1.HealthCheckRequest) (*grpc_health_v1.HealthCheckResponse, error) {
+	if g.code != codes.OK {
+		return nil, status.Error(g.code, r.GetService())
+	}
+
+	return &grpc_health_v1.HealthCheckResponse{Status: g.status}, nil
+}
+
+func (g *grpcServer) Watch(r *grpc_health_v1.HealthCheckRequest, s grpc_health_v1.Health_WatchServer) error {
+	panic("implement me")
```

Comment:
Slightly nicer style is to use `status.Errorf`, what do you think?
```suggestion
	return status.Error(codes.Unimplemented, "unimplemented")
```

Reply:
Nice one, I just kept the autogenerated code for this one, but looks cleaner like this 👍